cpp/src/arrow/compare.h

```diff
@@ -142,4 +152,208 @@ bool ARROW_EXPORT
+  using V = decltype(left.value);
+  Op<V> cmp;
+  if (left_nan) {
+    if (right_nan) {
+      return false;  // arbitrary - NaNs are unordered
+    } else {
+      return options.nans_least() ? cmp(true) : cmp(false);
+    }
+  } else {
+    if (right_nan) {
+      return options.nans_least() ? cmp(false) : cmp(true);
+    } else {
+      V tolerance = fp_type ? TypeAdjustedTolerance<V>(options.atolf())
+                            : TypeAdjustedTolerance<V>(options.atold());
+      return cmp(left.value, right.value, tolerance);
+    }
+  }
+}
+
+template <typename T>
+struct LessThanOp {
+  bool operator()(T left, T right, T tolerance) {
+    return (left < right) || (left - right < tolerance);
+  }
+  bool operator()(bool left_least) { return left_least; }
+};
+
+template <typename T>
+struct IsAtMostOp {
+  bool operator()(T left, T right, T tolerance) {
+    return (left <= right) || (left - right <= tolerance);
+  }
+  bool operator()(bool left_least) { return left_least; }
+};
+
+template <typename T>
+struct MoreThanOp {
+  bool operator()(T left, T right, T tolerance) {
+    return (left > right) || (right - left < tolerance);
+  }
+  bool operator()(bool left_least) { return !left_least; }
+};
+
+template <typename T>
+struct IsAtLeastOp {
+  bool operator()(T left, T right, T tolerance) {
+    return (left >= right) || (right - left <= tolerance);
+  }
+  bool operator()(bool left_least) { return !left_least; }
+};
+
+}  // namespace internal
+
+/// Returns true if left numeric scalar is less than right numeric scalar
+/// \param[in] left a NumericScalar
+/// \param[in] right a NumericScalar
+/// \param[in] options comparison options
+template <typename T>
+bool ScalarLessThan(const NumericScalar<T>& left, const NumericScalar<T>& right,
+                    const OrderOptions& options = OrderOptions::Defaults()) {
+  return internal::NumericScalarCompare<NumericScalar<T>, internal::LessThanOp>(
+      left, right, options);
+}
+
+/// Returns true if left numeric scalar is at most right numeric scalar
+/// \param[in] left a NumericScalar
+/// \param[in] right a NumericScalar
+/// \param[in] options comparison options
+template <typename T>
+bool ScalarIsAtMost(const NumericScalar<T>& left, const NumericScalar<T>& right,
+                    const OrderOptions& options = OrderOptions::Defaults()) {
+  return internal::NumericScalarCompare<NumericScalar<T>, internal::IsAtMostOp>(
+      left, right, options);
+}
+
+/// Returns true if left numeric scalar is more than right numeric scalar
+/// \param[in] left a NumericScalar
+/// \param[in] right a NumericScalar
+/// \param[in] options comparison options
+template <typename T>
+bool ScalarMoreThan(const NumericScalar<T>& left, const NumericScalar<T>& right,
+                    const OrderOptions& options = OrderOptions::Defaults()) {
+  return internal::NumericScalarCompare<NumericScalar<T>, internal::MoreThanOp>(
+      left, right, options);
```

Comment:
Could perhaps save on code generation by avoiding a separate op for this:
```suggestion
  return internal::NumericScalarCompare<NumericScalar<T>, internal::LessEqualOp>(
      right, left, options);
```